### PR TITLE
AppAction interface for BDF2

### PR DESCRIPTION
**Pull request by asteyer**

### CODE REVIEW

**Comment by asteyer** [comment]
@ccober6 These are the actual BDF2 AppAction changes - the other PR got screwed up when I squashed and didn't add any of the changes.

File: packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+ *  \renewcommand{\thealgorithm}{}
+ *  \caption{BDF2 with the locations of the application actions indicated}
+ *  \begin{algorithmic}[1]
+ *    \State Start with $x_n$, $\Delta t_n$
+ *    \State {\it appAction.execute(solutionHistory, stepper, BEGIN\_STEP)}
+ *    \State Form $f(x_{n},t_{n})$
+ *    \State {\it appAction.execute(solutionHistory, stepper, BEFORE\_EXPLICIT\_EVAL)}
+ *    \State Form $x_n \leftarrow x_{n} + \Delta t_n f(x_{n},t_n)$
+ *    \State {\it appAction.execute(solutionHistory, stepper, END\_STEP)}
+ *  \end{algorithmic}
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:44]: Need to make this match the algorithm: BEGIN_STEP, BEFORE_SOLVE, AFTER_SOLVE, and END_STEP.

File: packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      case StepperBDF2AppAction<Scalar>::AFTER_SOLVE:
+      {
+        modType = X_AFTER_SOLVE;
+        x = workingState->getX();
+        break;
+      }
+      case StepperBDF2AppAction<Scalar>::END_STEP:
+      {
+        modType = XDOT_END_STEP;
+        x = stepper->getStepperXDot();
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:102]: We should make this X_END_STEP and not XDOT_END_STEP, using getStepperX().

File: packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+public:
+
+  /// Indicates the location of application action (see algorithm).
+  enum MODIFIER_TYPE {
+    X_BEGIN_STEP,     ///< Modify \f$x\f$ at the beginning of the step.
+    X_BEFORE_SOLVE,   ///< Modify \f$x\f$ before the implicit solve.
+    X_AFTER_SOLVE,    ///< Modify \f$x\f$ after the implicit solve
+    XDOT_END_STEP     ///< Modify \f$\dot{x}\f$ at the end of the step.
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:120]: Same here.

File: packages/tempus/src/Tempus_StepperBDF2ModifierXDefault.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  virtual void modify(
+    Teuchos::RCP<Thyra::VectorBase<Scalar> > /* x */,
+    const Scalar /* time */, const Scalar /* dt */,
+    const typename StepperBDF2ModifierXBase<Scalar>::MODIFIER_TYPE modType)
+  {
+    switch(modType) {
+      case StepperBDF2ModifierXBase<Scalar>::X_BEGIN_STEP:
+      case StepperBDF2ModifierXBase<Scalar>::X_BEFORE_SOLVE:
+      case StepperBDF2ModifierXBase<Scalar>::X_AFTER_SOLVE:
+      case StepperBDF2ModifierXBase<Scalar>::XDOT_END_STEP:
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ModifierXDefault.hpp:46]: Same here.

File: packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  /// Observe BDF2 Stepper at end of takeStep.
+  virtual void observe(
+    Teuchos::RCP<const SolutionHistory<Scalar> > /* sh */,
+    Teuchos::RCP<const StepperBDF2<Scalar> > /* stepper */,
+    const typename StepperBDF2AppAction<Scalar>::ACTION_LOCATION actLoc)
+  {
+    switch(actLoc) {
+      case StepperBDF2AppAction<Scalar>::BEFORE_SOLVE:
+      case StepperBDF2AppAction<Scalar>::AFTER_SOLVE:
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp:44]: Need to add BEGIN_STEP and END_STEP.

File: packages/tempus/unit_test/Tempus_UnitTest_BDF2.cpp [hunk 6]
```diff
@@ -25,75 +30,472 @@
...
+        testDt = dt;
+        break;
+      }
+    case StepperBDF2ModifierXBase<double>::X_AFTER_SOLVE:
+      {
+        testX_AFTER_SOLVE = true;
+        testTime = time;
+        break;
+      }
+    case StepperBDF2ModifierXBase<double>::XDOT_END_STEP:
```
Comment (ccober6) [inline packages/tempus/unit_test/Tempus_UnitTest_BDF2.cpp:412]: Switch to X_END_STEP.

File: packages/tempus/unit_test/Tempus_UnitTest_BDF2.cpp [hunk 7]
```diff
@@ -25,75 +30,472 @@
...
+      }
+    default:
+      TEUCHOS_TEST_FOR_EXCEPTION(true, std::logic_error,
+                                 "Error - unknown action location.\n");
+    }
+  }
+  bool testX_BEGIN_STEP;
+  bool testX_BEFORE_SOLVE;
+  bool testX_AFTER_SOLVE;
+  bool testXDOT_END_STEP;
```
Comment (ccober6) [inline packages/tempus/unit_test/Tempus_UnitTest_BDF2.cpp:426]: Same here.

**Review by ccober6 — CHANGES_REQUESTED** [review]
Just a few mods.

File: packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp (hunk 1 above)
Comment (asteyer) [reply packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:44]: @ccober6 So no get Xdots anywhere?

File: packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp (hunk 2 above)
Comment (asteyer) [reply packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:102]: @ccober6 So no getXdots anywhere for BDF2?

Comment (ccober6) [reply packages/tempus/src/Tempus_StepperBDF2ModifierXBase.hpp:102]: Yeah, in talking with the EMPIRE folks they leaned towards just accessing X and not XDot.

**Review by ccober6 — APPROVED** [review]
Looks good!

**Comment by ccober6** [comment]
@asteyer So I did a checkout, build and run of branch 7530-AppAction-for-BDF2.  There seems to be a couple problems.  

First, I did not see the build errors that are appearing in this PR during autotesting.  I think the PR is in a bad state. 

@trilinos/framework can you help us out with this?  The build error appears is a previous problem that should be fixed now, but we are still seeing it during the auto-testing.

Second, my build gets the following tests failing

The following tests FAILED:
	  8 - Tempus_BDF2_MPI_1 (Failed)
	  9 - Tempus_BDF2_Combined_FSA_MPI_1 (Failed)
	 10 - Tempus_BDF2_Staggered_FSA_MPI_1 (Failed)
	 11 - Tempus_BDF2_PseudoTransient_SA_MPI_1 (Failed)
	 12 - Tempus_BDF2_ASA_MPI_1 (Failed)

@asteyer can you double check that you have pushed everything?  Also, can you do a clean checkout, build and run of this PR, and see if you get the same failures.

**Comment by jwillenbring** [comment]
> @trilinos/framework can you help us out with this? The build error appears is a previous problem that should be fixed now, but we are still seeing it during the auto-testing.

@ccober6 - Where we can, we reuse build trees between tests. This typically works well, but there are instances where it can cause problems. Once a week, we start with all new build trees across all machines for PR testing. I have manually invoked this cleaning step, so the next time this test is run, a full rebuild will be triggered.

**Comment by ccober6** [comment]
Thanks @jwillenbring !

**Comment by ccober6** [comment]
@asteyer Please let me know if you are working on this, as I am starting to work on fixes.

**Comment by ccober6** [comment]
@asteyer Fixed the bug in BDF2 that was causing failures, but also did a forced push.  So your branch is now likely broken.  Just throw away that branch, do a pull, and checkout 7530-AppAction-for-BDF2.

**Comment by ccober6** [comment]
@sconde Can you review this now?  Both @asteyer and myself have made edits, so we need a third party! :)

**Comment by asteyer** [comment]
@ccober6 Can you approve the changes here so I can merge and close this PR?

**Comment by ccober6** [comment]
@asteyer Since I have made edits, we should someone else to review it.  I have asked @sconde if he could review it for us.

**Comment by ccober6** [comment]
@sconde Can you review this for us? Thanks Curt.

**Comment by sconde** [comment]
on it

File: packages/tempus/src/Tempus_StepperBDF2AppActionCompsite.hpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "Tempus_StepperBDF2AppAction.hpp"
+#include "Tempus_TimeStepControl.hpp"
+#include <vector>
+
+namespace Tempus {
+
+/** \brief This composite AppAction loops over added AppActions.
+ *
+ *  Inidividual AppActions are executed in the order in which they
```
Comment (sconde) [inline packages/tempus/src/Tempus_StepperBDF2AppActionCompsite.hpp:20]: Individual - word misspelled.

**Review by sconde — DISMISSED** [review]
Looks good to me.

**Comment by ccober6** [comment]
@sconde So some how this PR got closed but not merged?!  I reopened it, rebased it and fixed some minor conflicts.  Can you review/approve this?

**Review by sconde — APPROVED** [review]

**Pull request merged**